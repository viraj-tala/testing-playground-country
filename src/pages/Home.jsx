import React, { useEffect, useState } from 'react';
import SearchInput from '../component/SearchComponent';
import CountriesTable from '../component/CountryTable';

let searchTimeout = null
const Home = () => {
    const [keyword, setKeyword] = useState('');
    const [countries, setCountries] = useState([]);

    useEffect(() => {
        const getCountries = async () => {
            try {
                const res = await fetch("https://restcountries.com/v3.1/all");
                const countryData = await res.json();
                if (countryData.length) setCountries(countryData);
            } catch (error) {
                console.log('Error fetching country data:', error);
            }
        };
        getCountries();
    }, []);

    useEffect(() => {
        return () => {
            if (searchTimeout) clearTimeout(searchTimeout)
        }
    }, [])

    const filteredCountries = countries.filter(
        ({ name, region, subregion }) => (
            name?.common?.toLowerCase().includes(keyword.toLowerCase()) ||
            region?.toLowerCase().includes(keyword.toLowerCase()) ||
            subregion?.toLowerCase().includes(keyword.toLowerCase())
        )
    )

    const onInputChange = (event) => {
        if (!event.target.value) {
            return
        }
        if (searchTimeout) clearTimeout(searchTimeout)
        searchTimeout = setTimeout(() => setKeyword(event.target.value), 500)
    };

    return (
        <>
            <div className="h-full w-full flex flex-col overflow-hidden">
                <div className="text-center flex flex-col mt-8 text-xs">
                    <h1 className="mb-4 text-4xl font-extrabold leading-none tracking-tight text-gray-900 md:text-5xl lg:text-6xl">Found<span className="text-blue-600 dark:text-blue-500"> {countries.length}</span> countries</h1>
                    <div className='flex items-center flex-col justify-between m-2'>
                        <SearchInput placeholder="Filter by Name, Region, or SubRegion" onChange={onInputChange} />
                    </div>
                </div>

                <main className="overflow-auto">
                    <CountriesTable countries={filteredCountries} />
                </main>

                <footer className="text-center mt-8 text-xs"></footer>
            </div >
        </>
    );
};

export default Home;
