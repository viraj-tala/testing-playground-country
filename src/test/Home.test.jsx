/* eslint-disable testing-library/no-render-in-setup */
/* eslint-disable testing-library/render-result-naming-convention */
import React from 'react';
import { render, screen } from './test.utils';
import Home from '../pages/Home';
import { CloudCircle } from '@material-ui/icons';

// Mocking the fetch function
jest.mock('../loader/Loader');

// Test suite for the Home component
describe('Home Component', () => {

    beforeEach(() => {
        render(
            <Home />
        );
    });

    it('should match the snapshot', () => {

        const element = render(<Home />)

        expect(element).toMatchSnapshot();
    })

    it('renders correctly', async () => {

        // Wait for the text 'Found' to be present in the document
        await screen.findByText(/Found/i);

        // Check if certain elements are present in the document
        expect(screen.getByText(/Found/i)).toBeInTheDocument();
        expect(screen.getByPlaceholderText('Filter by Name, Region, or SubRegion')).toBeInTheDocument();
    });
});
