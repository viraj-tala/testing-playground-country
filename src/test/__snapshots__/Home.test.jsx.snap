// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Component should match the snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="h-full w-full flex flex-col overflow-hidden"
      >
        <div
          class="text-center flex flex-col mt-8 text-xs"
        >
          <h1
            class="mb-4 text-4xl font-extrabold leading-none tracking-tight text-gray-900 md:text-5xl lg:text-6xl"
          >
            Found
            <span
              class="text-blue-600 dark:text-blue-500"
            >
               
              0
            </span>
             countries
          </h1>
          <div
            class="flex items-center flex-col justify-between m-2"
          >
            <div>
              <form>
                <div
                  class="m-4 relative "
                >
                  <div
                    class="absolute inset-y-0 start-0 flex items-center ps-3 pointer-events-none"
                  >
                    <svg
                      aria-hidden="true"
                      class="w-4 h-4 text-gray-500 dark:text-gray-400"
                      fill="none"
                      viewBox="0 0 20 20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="m19 19-4-4m0-7A7 7 0 1 1 1 8a7 7 0 0 1 14 0Z"
                        stroke="currentColor"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                        stroke-width="2"
                      />
                    </svg>
                  </div>
                  <input
                    class="block w-full p-4 ps-10 text-sm text-gray-900 border border-gray-300 rounded-lg bg-gray-50 focus:ring-blue-500 focus:border-blue-500 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500"
                    placeholder="Filter by Name, Region, or SubRegion"
                  />
                </div>
              </form>
            </div>
          </div>
        </div>
        <main
          class="overflow-auto"
        >
          <div />
        </main>
        <footer
          class="text-center mt-8 text-xs"
        />
      </div>
    </div>
    <div>
      <div
        class="h-full w-full flex flex-col overflow-hidden"
      >
        <div
          class="text-center flex flex-col mt-8 text-xs"
        >
          <h1
            class="mb-4 text-4xl font-extrabold leading-none tracking-tight text-gray-900 md:text-5xl lg:text-6xl"
          >
            Found
            <span
              class="text-blue-600 dark:text-blue-500"
            >
               
              0
            </span>
             countries
          </h1>
          <div
            class="flex items-center flex-col justify-between m-2"
          >
            <div>
              <form>
                <div
                  class="m-4 relative "
                >
                  <div
                    class="absolute inset-y-0 start-0 flex items-center ps-3 pointer-events-none"
                  >
                    <svg
                      aria-hidden="true"
                      class="w-4 h-4 text-gray-500 dark:text-gray-400"
                      fill="none"
                      viewBox="0 0 20 20"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="m19 19-4-4m0-7A7 7 0 1 1 1 8a7 7 0 0 1 14 0Z"
                        stroke="currentColor"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                        stroke-width="2"
                      />
                    </svg>
                  </div>
                  <input
                    class="block w-full p-4 ps-10 text-sm text-gray-900 border border-gray-300 rounded-lg bg-gray-50 focus:ring-blue-500 focus:border-blue-500 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500"
                    placeholder="Filter by Name, Region, or SubRegion"
                  />
                </div>
              </form>
            </div>
          </div>
        </div>
        <main
          class="overflow-auto"
        >
          <div />
        </main>
        <footer
          class="text-center mt-8 text-xs"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="h-full w-full flex flex-col overflow-hidden"
    >
      <div
        class="text-center flex flex-col mt-8 text-xs"
      >
        <h1
          class="mb-4 text-4xl font-extrabold leading-none tracking-tight text-gray-900 md:text-5xl lg:text-6xl"
        >
          Found
          <span
            class="text-blue-600 dark:text-blue-500"
          >
             
            0
          </span>
           countries
        </h1>
        <div
          class="flex items-center flex-col justify-between m-2"
        >
          <div>
            <form>
              <div
                class="m-4 relative "
              >
                <div
                  class="absolute inset-y-0 start-0 flex items-center ps-3 pointer-events-none"
                >
                  <svg
                    aria-hidden="true"
                    class="w-4 h-4 text-gray-500 dark:text-gray-400"
                    fill="none"
                    viewBox="0 0 20 20"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="m19 19-4-4m0-7A7 7 0 1 1 1 8a7 7 0 0 1 14 0Z"
                      stroke="currentColor"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                      stroke-width="2"
                    />
                  </svg>
                </div>
                <input
                  class="block w-full p-4 ps-10 text-sm text-gray-900 border border-gray-300 rounded-lg bg-gray-50 focus:ring-blue-500 focus:border-blue-500 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500"
                  placeholder="Filter by Name, Region, or SubRegion"
                />
              </div>
            </form>
          </div>
        </div>
      </div>
      <main
        class="overflow-auto"
      >
        <div />
      </main>
      <footer
        class="text-center mt-8 text-xs"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
